### PR TITLE
[Dev] Fix Linear-Cross-Entropy Convergence Issue

### PR DESCRIPTION
**Pull request by shjwudp**
# What does this PR do ?

main branch: https://github.com/NVIDIA/Megatron-LM/pull/3076
NOTE: The kernel will be migrated to TE in future updates, and the main branch will depend on this version of the kernel.

## 1. Fix: incorrect parameter reference in Linear-Cross-Entropy Fusion for non-shared embeddings

When using non-shared embeddings, `Linear-Cross-Entropy Fusion` incorrectly references the shared embedding weight. In this scenario, the return value of `shared_embedding_or_output_weight` should not be used. Instead, it should explicitly reference `output_layer.weight`.

Additionally, since `Linear-Cross-Entropy Fusion` does not invoke the `output_layer` module, m-core DDP’s parameter gathering check may fail. To address this, I subclassed `output_layer` (i.e., `ColumnParallelLinear`) and moved the Linear-Cross-Entropy logic into its `forward` method.

## 2. Improvement: Use FP32 accumulation in multi-chunk computation for better convergence

The previous implementation of Linear-Cross-Entropy Fusion split computation along the vocabulary dimension and accumulated gradients in BF16, using the operation:
```
d_hidden += d_logits_slice * weight_slice
```

This caused `d_hidden` to accumulate in BF16, which could lead to reduced numerical accuracy.

This PR fixes the issue by allocating an FP32 buffer for `d_hidden` during accumulation, then downcasting it back to BF16 after accumulation completes. This improves numerical stability and overall convergence.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by shjwudp** [comment]
Hi @NVIDIA/core-devtech @NVIDIA/core-nemo , could you help review this PR? Jianbing has reviewed and approved the changes to the kernel.

**Comment by yanring** [comment]
Hi Jack, could you please add the link to the main PR in the desc? Thanks.

**Comment by ethen8181** [comment]
Hi team, wanted to see whether the linear-cross-entropy fused kernel is ready for use. Or if there's a central place where we can follow to check when it will be released in the main branch, thanks
